torch/csrc/jit/tracer.h

```diff
@@ -0,0 +1,219 @@
+#pragma once
+
+#include "torch/csrc/jit/ir.h"
+#include "torch/csrc/jit/assert.h"
+#include "torch/csrc/autograd/function_hook.h"
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/jit/init_pass.h"
+
+#include <memory>
+#include <mutex>
+#include <vector>
+#include <iostream>
+#include <cstdint>
+#include <unordered_map>
+
+
+namespace torch { namespace jit { namespace tracer {
+
+using torch::autograd::Variable;
+using variable_list = std::vector<std::shared_ptr<Variable>>;
+
+// TracingState tracks the necessary state when we are tracing the execution of
+// autograd code; most importantly, it holds a reference to the actual IR
+// graph which we are recording the trace to.
+//
+// The liveness of a TracingState is expected to be a superset of the region
+// of code being traced; in particular, Variables do not keep a TracingState
+// live.  Instead, they hold weak pointers to TracingState, to prevent leaks
+// from arising when a variable that participated in a trace outlives the
+// actual trace itself.
+//
+// Multi-threaded tracing is NOT yet supported: it is better to think of
+// tracing as a thread local affair, where we enter and then exit
+// a tracing region.  This not only coincides with how Python code
+// is run (GIL = single-threaded), but also how single Functions in
+// an autograd closure are applied.  Note that the execution of an
+// *entire* autograd closure is multithreaded, in which case extra
+// locking is necessary.
+
+struct TracingState : public std::enable_shared_from_this<TracingState> {
+  TracingState()
+    : graph(new Graph())
+    , active(false) {}
+
+  std::shared_ptr<Graph> graph;
+  // void* is an unsafe TH.  NON-OWNING, so it might get invalidated.
+  // TODO: Perhaps, turn this into an owning reference.  The buffers
+  // are persistent, so this won't lead to a leak.
+  std::unordered_map<void*, Node*> buffer_map;
```

Comment:
Why does it need to be part of the tracing state? We could only give it to the exporter, right?

Reply:
The best time to initialize the buffer_map is when we initiate the trace (since we get in the buffers from the state_dict.) But then, the most convenient way to pass this along to the exporter is to put it in the tracing state.